src/nexmark/queries/q4.rs

```diff
@@ -0,0 +1,208 @@
+use super::NexmarkStream;
+use crate::{nexmark::model::Event, operator::FilterMap, Circuit, OrdZSet, Stream};
+
+/// Query 4: Average Price for a Category
+///
+/// Select the average of the wining bid prices for all auctions in each
+/// category. Illustrates complex join and aggregation.
+///
+/// From [Nexmark q4.sql](https://github.com/nexmark/nexmark/blob/v0.2.0/nexmark-flink/src/main/resources/queries/q4.sql):
+///
+/// ```sql
+/// DROP VIEW IF EXISTS auction;
+/// DROP VIEW IF EXISTS bid;
+/// CREATE VIEW auction AS SELECT auction.* FROM ${NEXMARK_TABLE} WHERE event_type = 1;
+/// CREATE VIEW bid AS SELECT bid.* FROM ${NEXMARK_TABLE} WHERE event_type = 2;
+///
+/// CREATE TABLE discard_sink (
+///   id BIGINT,
+///   final BIGINT
+/// ) WITH (
+///   'connector' = 'blackhole'
+/// );
+///
+/// INSERT INTO discard_sink
+/// SELECT
+///     Q.category,
+///     AVG(Q.final)
+/// FROM (
+///     SELECT MAX(B.price) AS final, A.category
+///     FROM auction A, bid B
+///     WHERE A.id = B.auction AND B.dateTime BETWEEN A.dateTime AND A.expires
+///     GROUP BY A.id, A.category
+/// ) Q
+/// GROUP BY Q.category;
+/// ```
+
+type Q4Stream = Stream<Circuit<()>, OrdZSet<(usize, usize), isize>>;
+
+pub fn q4(input: NexmarkStream) -> Q4Stream {
+    // Select auctions and index by auction id.
+    let auctions_by_id = input.flat_map_index(|event| match event {
+        Event::Auction(a) => Some((a.id, (a.category, a.date_time, a.expires))),
+        _ => None,
+    });
+
+    // Select bids and index by auction id.
+    let bids_by_auction = input.flat_map_index(|event| match event {
+        Event::Bid(b) => Some((b.auction, (b.price, b.date_time))),
+        _ => None,
+    });
+
+    type BidsAuctionsJoin =
+        Stream<Circuit<()>, OrdZSet<((u64, usize, u64, u64), (usize, u64)), isize>>;
+
+    // Join to get bids for each auction.
+    let bids_for_auctions: BidsAuctionsJoin = auctions_by_id.join::<(), _, _, _>(
+        &bids_by_auction,
+        |&auction_id, &(category, a_date_time, a_expires), &(bid_price, bid_date_time)| {
+            (
+                (auction_id, category, a_date_time, a_expires),
+                (bid_price, bid_date_time),
+            )
+        },
+    );
+
+    // Filter out the invalid bids while indexing.
+    let bids_for_auctions_indexed = bids_for_auctions.flat_map_index(
+        |&((auction_id, category, a_date_time, a_expires), (bid_price, bid_date_time))| {
+            if bid_date_time >= a_date_time && bid_date_time <= a_expires {
+                Some(((auction_id, category), bid_price))
+            } else {
+                None
+            }
+        },
+    );
+
+    // winning_bids_by_category: once we have the winning bids, we don't
+    // need the auction ids anymore.
+    let winning_bids_by_category: Stream<Circuit<()>, OrdZSet<(usize, usize), isize>> =
+        bids_for_auctions_indexed.aggregate(|&key, vals| -> (usize, usize) {
```

Comment:
We must use incremental version of `aggregate` here for correctness: since all previous operators are incremental, inputs to `aggregate` are changes to the `bids_for_auctions_indexed` relation.  The non-incremental `aggregate` will aggregate values in the last delta only.

Reply:
OK, I'm going to try to update my test to fail to better understand this. Thanks.

Reply:
PTAL. I've updated to use `aggregate_incremental` in both instances. I didn't spend too long on it, but wasn't able to make my test fail only for the non-incremental version: Is the last delta calculated for each batch? I tried adding an extra `root.step.unwrap`, but then it fails both for the incremental and non-incremental - which surprised me, I assumed that the next step would just not have any deltas and so the expected value would not have changed.

Reply:
You do need more than one iteration to observe the difference between incremental and non-incremental implementation. Given a query `q` and an input `i` at time 0, the non-incremental implementation computes `q(i)`, while the incremental one computes `q(i) - q(0)`, and since for this query`q(0) = 0`, the two outputs are the same.

Furthermore, you need the second iteration of your terst to take a non-empty input.  If your first input is `i` and your second input is `0`, then the non-incremental query computes `q(i)` at the first iteration and `q(0) = 0` at the second iteration.  The incremental query computes `q(i) - q(0) = q(i)` at the first iteration and `q(i+0) - q(i) = q(i) - q(i) = 0` at the second iteration.  In other words, if your input doesn't change then neither does the output.

> I assumed that the next step would just not have any deltas and so the expected value would not have changed.

Right, and for the incremental implementation this means outputting an empty batch at the second step.